File: skellig-test-step-processing/src/main/java/org/skellig/teststep/processing/validation/comparator/ContainsValueComparator.java
package org.skellig.teststep.processing.validation.comparator;

import java.util.Arrays;
import java.util.Collection;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

class ContainsValueComparator implements ValueComparator {

    private static final Pattern CONTAINS_PATTERN = Pattern.compile("contains\\((.+)\\)");

    @Override
    public boolean compare(Object actualValue, Object expectedValue) {
        if (actualValue != null) {
            Matcher matcher = CONTAINS_PATTERN.matcher(String.valueOf(expectedValue));
            if (matcher.find()) {
                String expectedValueAsString = matcher.group(1);
                // usually actual is String so to speed up comparison it checks if it is String first
                if (actualValue.getClass().equals(String.class)) {
                    return ((String) actualValue).contains(expectedValueAsString);
                } else if (actualValue.getClass().isArray()) {
                    return Arrays.stream((Object[]) actualValue)
                            .map(String::valueOf)
                            .anyMatch(item -> item.equals(expectedValueAsString));
                } else if (actualValue instanceof Collection) {
                    return ((Collection<Object>) actualValue).stream()
                            .map(String::valueOf)
                            .anyMatch(item -> item.equals(expectedValueAsString));
                } else {
                    return String.valueOf(actualValue).contains(expectedValueAsString);
                }
            }
        }
        return false;
    }

    @Override
    public boolean isApplicable(Object expectedValue) {
        return CONTAINS_PATTERN.matcher(String.valueOf(expectedValue)).matches();
    }
}

File: skellig-test-step-processing/src/main/kotlin/org/skellig/teststep/processing/validation/comparator/ContainsValueComparator.kt
package org.skellig.teststep.processing.validation.comparator

class ContainsValueComparator : ValueComparator {

    override fun compare(comparator: String, args: Array<Any?>, actualValue: Any?): Boolean {
        if (args.isNotEmpty() && actualValue != null) {
            return args.filterNotNull().map { it.toString() }.all {
                when {
                    actualValue::class == String::class -> (actualValue as String).contains(it)
                    actualValue.javaClass.isArray -> compareArray(actualValue, it)
                    actualValue is Collection<*> -> compareCollection(actualValue, it)
                    actualValue is Map<*, *> -> compareMap(actualValue, it)
                    else -> actualValue.toString().contains(it)
                }
            }
        }
        return false
    }

    private fun compareCollection(actualValue: Collection<*>, expectedValueAsString: String?) = actualValue
        .map { it.toString() }
        .any { it == expectedValueAsString }

    private fun compareMap(actualValue: Map<*, *>, expectedValueAsString: String?) =
        compareCollection(actualValue.values, expectedValueAsString)

    private fun compareArray(actualValue: Any?, expectedValueAsString: String?) =
        if (actualValue is ByteArray) {
            actualValue
                .map { it.toString() }
                .any { it == expectedValueAsString }
        } else {
            (actualValue as Array<*>)
                .map { it.toString() }
                .any { it == expectedValueAsString }
        }

    override fun getName(): String = "contains"
}